tools/buildsys/src/gomod.rs

```diff
@@ -0,0 +1,176 @@
+/*!
+Packages using the Go programming language may have upstream tar archives that
+include only the source code of the project, but not the source code of any
+dependencies. The Go programming language promotes the use of "modules" for
+dependencies and projects adopting modules will provide go.mod and go.sum
+files.
+
+This module extends the functionality of `packages.metadata.build-package.external-files`
+and provides the ability to retrieve and validate dependencies
+declared using Go modules given a tar archive containing a go.mod and go.sum.
+
+The location where dependencies are retrieved from are controlled by the
+standard environment variables employed by the Go tool: GOPROXY, GOSUMDB, and
+GOPRIVATE.
+
+ */
+
+pub(crate) mod error;
+use error::Result;
+
+use super::manifest;
+use duct::cmd;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::env;
+use std::path::{Path, PathBuf};
+use std::process::Output;
+
+pub(crate) struct GoMod;
+
+impl GoMod {
+    pub(crate) fn vendor(
+        root_dir: &Path,
+        package_dir: &Path,
+        gomod: &manifest::ExternalFile,
+    ) -> Result<()> {
+        let url_file_name = extract_file_name(&gomod.url)?;
+        let path_arg = &gomod.path.as_ref().unwrap_or(&url_file_name);
+        ensure!(path_arg.components().count() == 1, error::InputFileSnafu);
+
+        let path = package_dir.join(path_arg);
+        ensure!(path.is_file(), error::InputFileBadSnafu { path });
+        let mod_dir = gomod.bundle_path.as_ref().context(error::ModDirSnafu)?;
+        let output_path_arg = gomod
+            .bundle_output_path
+            .as_ref()
+            .context(error::OutputDirSnafu)?;
+        let output_path = package_dir.join(output_path_arg);
+
+        // Return early if the output path exists and is a file, assuming it's already been built
+        if output_path.exists() && output_path.is_file() {
+            return Ok(());
+        }
+
+        // Our SDK and toolchain are picked by the external `cargo make` invocation.
+        let sdk = getenv("BUILDSYS_SDK_IMAGE")?;
+
+        // Several Go variables control proxying
+        let mut goproxy = go_env("GOPROXY").unwrap_or_else(|| "".to_string());
+        if goproxy.ends_with('\n') {
+            goproxy.pop();
+        }
+        let mut gosumdb = go_env("GOSUMDB").unwrap_or_else(|| "".to_string());
+        if gosumdb.ends_with('\n') {
+            gosumdb.pop();
+        }
+        let mut goprivate = go_env("GOPRIVATE").unwrap_or_else(|| "".to_string());
+        if goprivate.ends_with('\n') {
+            goprivate.pop();
+        }
```

Comment:
nit: maybe factor this into a helper function that takes a list of possible environment variables

Reply:
This code block was redundant since `docker-go` gets these variables from the environment so this block was removed.